src/shaders/heatmap.fragment.glsl

```diff
@@ -0,0 +1,21 @@
+#pragma mapbox: define highp float weight
+
+uniform float u_intensity;
+varying vec2 v_extrude;
+
+// Gaussian kernel coefficient: 1 / sqrt(2 * PI)
+#define GAUSS_COEF 0.3989422804014327
+
+void main() {
+    #pragma mapbox: initialize highp float weight
+
+    // Kernel density estimation with a Gaussian kernel of size 5x5
+    float d = -0.5 * 5.0 * 5.0 * dot(v_extrude, v_extrude);
+    float val = weight * u_intensity * GAUSS_COEF * exp(d);
```

Comment:
I don't think the 5^2 factor here technically translates to a kernel size of 5x5.  Here's my read/analysis to make sure I'm understanding this right:

The "bandwidth" of the Gaussian kernel (which is also its standard deviation) would be the value `h` in `1/2pi * exp( -0.5 * (u/h)^2 )`

Since `v_extrude` is essentially in units of radii (i.e. at a distance of `u_radius` from the point, `v_extrude` is a unit vector), the calculation above is equivalent to `weight * u_intensity * GAUSS_COEF * exp( (distance_in_pixels / (radius / 5)^ )`, making the bandwidth `radius/5`.

Three standard deviations is a pretty reasonable definition for the practical "size" of a Gaussian kernel.  In this case, that would be: `3/5 * radius`.

Reply:
Yeah, I noted that "radius" is quite a blurry term when referring to kernel density Gaussian kernels, but haven't thought this through further. Given the thinking above, how do you think we should change the formula in the shader?

Reply:
I think changing `5.0` to `3.0` would be pretty reasonable.  Then `radius` would have a pretty clean meaning:
 - For advanced users familiar w/ the underlying math, "3 standard deviations of the Gaussian kernel"
 - For most users, "the distance around each point where it influences the heatmap" (or something like that)

Reply:
Hmm, experimenting with this a bit, one issue is that with if we use `3.0`, then for values of `intensity * weight > 0.5`, the output for a point ends up getting clipped by the square mesh for that point.  Whereas with `5.0`, it takes `intensity * weight > 5000` to make that happen.  

I guess the problem is that, even at x = 3 standard deviations, the gaussian curve evaluates to `0.0018`, which is just half of the 1/255 resolution of the color ramp.

Reply:
Yeah, now I remember that empirically 5.0 is reasonable enough to make clipping artifacts not to be noticeable. Picking more, the performance will suffer because of the number of fragments. Any ideas on how to do better here?

Reply:
Do you think it'd be reasonable to restrict `heatmap-weight` to values between 0 and 1?  If so, then we could choose kernel bandwidth based on the intensity in such a way as to ensure that the kernel output drops to below `1/255/2` at ~ 1 radius:

```
float ZERO = 1.0/255.0/2.0;
float h = sqrt(-2.0 * log(ZERO / u_intensity / GAUSS_COEF));
float d = -0.5 * h * h * dot(v_extrude, v_extrude);
float val = weight * u_intensity * GAUSS_COEF * exp(d);
```

If `weight` can be arbitrarily large, though, then there's sort of no way around the problem (regardless of what value we choose, 3, 5, etc.)

Reply:
^ btw that `h` calculation need not happen in the shader -- could also be a uniform itself.  (Dunno if WebGL/OpenGL would optimize such a computation since it only relies on uniform values.)

Reply:
We can't restrict the weight, otherwise you won't be able to use clustering with heatmaps. If there's a cluster with 1000 points, it's weight should be 1000 times the single point weight.

Reply:
Other option would be to pick another type of kernel (not Gaussian) for the heatmap, or modify the Gaussian to fade out to 0 at the borders regardless of the weight.

Reply:
Trie Epanechnikov and the heatmap looks pretty bad. Maybe we could both calculate bandwidth and increase the size of the kernel depending on weight in the vertex shader? That way small intensities wouldn't require as many fragments, which would increase performance, and big intensities which cause big kernels wouldn't appear too often.

Reply:
@mourner Actually if we use a `weight`-based adjustment to the size of the kernel (i.e., the size of the mesh produced by the vertex shader), then we could go back to just using a simple, fixed value of `3.0` for the bandwidth.  I.e., in the vertex shader,

```glsl
    v_extrude = vec2(mod(a_pos, 2.0) * 2.0 - 1.0) * scale_based_on_weight;
```

And then in fragment

```glsl
       float d = -0.5 * h * h * dot(v_extrude / scale_based_on_weight, v_extrude / scale_based_on_weight);
```

Where `scale_based_on_weight` is a scale factor that grows the kernel large enough beyond `radius` so that the kernel falls effectively to zero at the edge

Reply:
@anandthakker exactly, that's what I was thinking too! Now we need to determine that scale properly...